File: src/com/mobiletheatertech/plot/TrussBase.java
package com.mobiletheatertech.plot;

import org.w3c.dom.Element;

/**
* Created with IntelliJ IDEA. User: dhs Date: 6/29/13 Time: 5:05 PM To change this template use
* File | Settings | File Templates.
 *
 * This is a square base for truss.
*
* @author dhs
* @since 0.0.5
*/
public class TrussBase extends MinderDom {

    private Double x;
    private Double y;
    private Double size;
    private Double rotation;

    private String color = "taupe";

    private String processedMark = null;


    public TrussBase(Element element ) throws AttributeMissingException, DataException, InvalidXMLException {
        super( element );

        size = getDoubleAttribute( "size" );
        y = getDoubleAttribute( "y" );
        x = getDoubleAttribute( "x" );
        rotation = getOptionalDoubleAttributeOrZero( "rotation" );

        // Give the base's element a unique id so that its Plot object can be recognized
        // and used.
        processedMark = Mark.Generate();
        element.setAttribute( "processedMark", processedMark );
    }

    /**
     * Find a {@code TrussBase}
     *
     * @param mark string to match while searching for a {@code TrussBase}
     * @return {@code TrussBase} whose mark matches specified string
     */
    // Copied from Truss - refactor to Minder?
    public static TrussBase Find(String mark ) {
        for (ElementalLister thingy : ElementalLister.List()) {
            if (TrussBase.class.isInstance( thingy )) {
                if (((TrussBase) thingy).processedMark.equals( mark )) {
                    return (TrussBase) thingy;
                }
            }
        }
        return null;
    }

//    public Point locate() {
//        return new Point( x, y, 0 );
//    }

    Double x() {
        return x;
    }

    Double y() {
        return y;
    }

    Double size() {
        return size;
    }

    Double rotation() {
        return rotation;
    }

    @Override
    public void verify() throws InvalidXMLException {
        //To change body of implemented methods use File | Settings | File Templates.
    }

    @Override
    public void dom( Draw draw, View mode ) {
        switch (mode) {
            case PLAN:
                SvgElement group = svgClassGroup( draw, Truss$.MODULE$.LayerTag() );
                draw.appendRootChild( group );
                SvgElement base = group.rectangle( draw, x - size / 2, y - size / 2, size, size, color );

                Double transformX = x + SvgElement.OffsetX();
                Double transformY = y + SvgElement.OffsetY();
                String transform = "rotate(" + rotation + "," + transformX + "," + transformY + ")";
                base.attribute( "transform", transform );
        }
    }
}

File: src/com/mobiletheatertech/plot/TrussBase.scala
package com.mobiletheatertech.plot

/**
  * Created by dhs on 11/29/16.
  */
import java.util

import org.w3c.dom.Element

/**
  * Created with IntelliJ IDEA. User: dhs Date: 6/29/13 Time: 5:05 PM To change this template use
  * File | Settings | File Templates.
  *
  * This is a square base for truss.
  *
  * @author dhs
  * @since 0.0.5
  */

class TrussBase (val element: Element) extends MinderDom(element)
  with Populate
{
  final val spacing = 3.0

  val size = getDoubleAttribute("size")
  val y = getDoubleAttribute("y")
  val x = getDoubleAttribute("x")
  val z = getOptionalDoubleAttributeOrZero( "z" )
  val rotation = getOptionalDoubleAttributeOrZero("rotation")
  var truss: Truss = null


  // Give the base's element a unique id so that its Plot object can be recognized
  // and used.
  val processedMark = Mark.Generate()
  element.setAttribute( "processedMark", processedMark )

  var drawPlace : Point = null

  tagCallback( Truss.LayerTag, processTruss )
  populate( element )


  def processTruss( element: Element ): Unit = {
    println( "TrussBase creates child Truss")
    truss = new Truss( element, this )
  }

  def verify() {
    drawPlace = Proscenium.LocateIfActive( new Point( x, y, z ) )
  }

  def mountPoints(): Array[Point] = {
    val foo = new Array[Point](4)
    foo( 0 ) = new Point( x - spacing, y - spacing, z + 2.0 )
    foo( 1 ) = new Point( x + spacing, y - spacing, z + 2.0 )
    foo( 2 ) = new Point( x - spacing, y + spacing, z + 2.0 )
    foo( 3 ) = new Point( x + spacing, y + spacing, z + 2.0 )
    foo
  }

  def dom(draw: Draw, mode: View) {
    mode match {
      case View.PLAN =>
        println( "TrussBase.dom." )
        val group = MinderDom.svgClassGroup(draw, Truss.LayerTag)
        draw.appendRootChild(group)

        val base = group.rectangle(draw, x - size / 2, y - size / 2, size, size, TrussBase.Color )

        val transformX: Double = x + SvgElement.OffsetX
        val transformY: Double = y + SvgElement.OffsetY
        val transform: String = "rotate(" + rotation + "," + transformX + "," + transformY + ")"
        base.attribute("transform", transform)

      case default =>
    }
  }
}

object TrussBase {
  final val Tag: String = "trussbase"
  final val Color = "taupe"

  /**
    * Find a {@code TrussBase}
    *
    * @param mark string to match while searching for a { @code TrussBase}
    * @return { @code TrussBase} whose mark matches specified string
    */
  // Copied from Truss - refactor to Minder?
  def Find(mark: String): TrussBase = {
    import scala.collection.JavaConversions._
//    for (thingy <- ElementalLister.List) {
//      if (classOf[TrussBase].isInstance(thingy)) if (thingy.asInstanceOf[TrussBase].processedMark == mark) return thingy.asInstanceOf[TrussBase]
//    }

    val list = ElementalLister.List()

    for( item <- 0 until list.size() ) {
      val thingy = list.get( item )
      if ( thingy.isInstanceOf[ TrussBase ]) {
        if ( thingy.asInstanceOf[ TrussBase ].processedMark == mark ) {
          return thingy.asInstanceOf[TrussBase]
        }
      }
    }

    null
  }

  def Reset(): Unit = {
//    LegendRegistered = false
//    LegendCount = 0
  }

}
